File: 01.DFS递归/leetcode_3336_subsequencePairCount.py
from typing import List
from functools import cache
from math import gcd

class Solution:
    def subsequencePairCount(self, nums: List[int]) -> int:
        n = len(nums)
        modNum = 1_000_000_007
        @cache
        def dfs(indx, g1, g2):
            if indx == n:
                if g1 == g2:
                    return 1
                return 0
            return (dfs(indx+1, gcd(g1, nums[indx]), g2)+dfs(indx+1, g1, gcd(g2, nums[indx]))+dfs(indx+1, g1, g2))%modNum
        return dfs(0, 0, 0)-1 # 减去「两个空序列相等」的情况

if __name__ == "__main__":
    s = Solution()
    nums = [1,2,3,4]
    print(s.subsequencePairCount(nums))

File: 01.DFS递归/leetcode_3336_subsequencePairCount.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <cmath>
#include <algorithm>
using namespace std;

class Solution {
private:
    unordered_map<int, int> memo;
    int n, modNum;
    vector<int> NUMS;

    int dfs(int indx, int g1, int g2){
        if (indx == n){
            if (g1 == g2){
                return 1;
            }
            return 0;
        }
        int key = (indx<<16) | (g1<<8) | g2;
        if (memo.count(key)){
            return memo[key];
        }
        return memo[key] = (1ll*dfs(indx+1, __gcd(g1, NUMS[indx]), g2)+
               1ll*dfs(indx+1, g1, __gcd(g2, NUMS[indx]))+
               1ll*dfs(indx+1, g1, g2))%modNum;
    }
public:
    int subsequencePairCount(vector<int>& nums) {
        modNum = 1e9+7;
        n = nums.size();
        NUMS = nums;
        return dfs(0, 0, 0)-1;
    }
};

int main(){
    Solution s;
    vector<int> nums = {1,2,3,4};
    cout << s.subsequencePairCount(nums) << endl;
}
